src/math/ec_scalar.cpp

```diff
@@ -18,36 +18,82 @@
  */
 #include <bitcoin/system/math/ec_scalar.hpp>
 
+#include <algorithm>
 #include <bitcoin/system/math/hash.hpp>
 #include <bitcoin/system/math/elliptic_curve.hpp>
+#include <bitcoin/system/utility/serializer.hpp>
 
 namespace libbitcoin {
 namespace system {
 
-// TODO: implement equality and inequality for ec_scalar and ec_secret.
 const ec_scalar ec_scalar::zero = null_hash;
 
-// Null is valid but uninitialized object is invalid.
 ec_scalar::ec_scalar()
-  : valid_(false), secret_(null_hash)
 {
+    // Avoid initializing secret by default.
+    // Helps when you pre-allocate large arrays of values.
+}
+
+ec_scalar::ec_scalar(uint64_t value)
+{
+    *this = value;
 }
 
 ec_scalar::ec_scalar(const ec_secret& secret)
-  : valid_(true), secret_(secret)
 {
+    *this = secret;
+}
+
+ec_scalar::ec_scalar(const ec_scalar& scalar)
+{
+    *this = scalar;
+}
+
+ec_scalar::ec_scalar(ec_scalar&& scalar)
+{
+    *this = std::move(scalar);
 }
 
 // Operators.
 // ----------------------------------------------------------------------------
 
+ec_scalar& ec_scalar::operator=(uint64_t value)
+{
+    secret_.reset(new ec_secret);
+    BITCOIN_ASSERT(secret_->size() == 32);
+    auto serial = bc::system::make_unsafe_serializer(secret_->begin());
+    serial.write_8_bytes_big_endian(0);
+    serial.write_8_bytes_big_endian(0);
+    serial.write_8_bytes_big_endian(0);
```

Comment:
Is there a 16 byte writer?

Reply:
You can probably use partial aggregate initialization here to fill the zeroes. You could also use it for the decoded int, but prob simpler to use the writer for that.

Reply:
You could also use our zeroize function to full the zeroes. Much faster than writing, maybe not faster than aggregate init.

Reply:
Fixed with `std::fill`

Reply:
Done